[cp] dispatch flex_attention to CP impl in TorchDispatchMode

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #155059
* #154691
* #152311
* __->__ #151497

## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

XilunWu committed:
DTensor HOP dispatch

[ghstack-poisoned]

XilunWu committed:
Update on "DTensor HOP dispatch"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "DTensor HOP dispatch"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "DTensor HOP dispatch"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "DTensor HOP dispatch"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

bdhirsh commented:
With https://github.com/pytorch/pytorch/pull/151719 (and a few changes to DTensor's `flex_attention` handling see here: P1789852198), I could run the E2E ring attention tests locally!

XilunWu committed:
Update on "DTensor HOP dispatch"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu committed:
Update

[ghstack-poisoned]

XilunWu committed:
Update

[ghstack-poisoned]

XilunWu committed:
Update

[ghstack-poisoned]

XilunWu committed:
Update on "[cp] dispatch flex_attention to CP impl in TorchDispatchMode"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[cp] dispatch flex_attention to CP impl in TorchDispatchMode"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[cp] dispatch flex_attention to CP impl in TorchDispatchMode"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+
+
+# TODO: merge with class `_LoadBalancer`
+class FlexAttentionSharder(ABC):
```
Make this a private class until we finalize the design.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
 _cp_options = _ContextParallelOptions()
 
 
+class _LoadBalancer(ABC):
```
Let's rename it to `_Sharder`. I'm not sure why I chose this bad name, lol.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+    ) -> torch.Tensor: ...
+
+
+class FlexAttentionContiguousSharder(FlexAttentionSharder):
```
ditto, make it a private class.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -1456,3 +1574,162 @@ def set_rotate_method(rotate_method: str) -> None:
+
+
+# NOTE: functools.lru_cache may cause memory leak
+@lru_cache  # noqa: B019
```
We should set a limit. `block_mask` contains tensors, which will cause OOM.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -1229,110 +1409,32 @@ def __torch_function__(
             with _enable_cp_dispatcher():
                 yield
     else:
```
`elif _dispatch_mode == _DispatchMode.TORCH_DISPATCH`. We should keep the `else` to handle the incorrect setting case.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+        tensor: torch.Tensor,
+        num_chunks: int,
+        shard_dim: int,
+        **kwargs: dict[str, Any],
```
This is not used in either flex attention sharder, can we remove this first? We can add this back later if we actually need it. We probably need tokens, but that is generic even for SDPA.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+    @abstractmethod
+    def shard(
+        self,
+        tensor: torch.Tensor,
```
Use `buffer` to make it consistent with the other sharders or if you want, you can change others to `tensor.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+    def shard(
+        self,
+        tensor: torch.Tensor,
+        num_chunks: int,
```
This number can easily infer from `device_mesh`. This way you can let FlexAttentionSharder inherit from `_LoadBalancer` or `_Sharder`.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+
+    @abstractmethod
+    def unshard(
+        self, shards: list[torch.Tensor], shard_dim: int, **kwargs: dict[str, Any]
```
ditto, remove `kwargs`.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+    ) -> torch.Tensor: ...
+
+    @abstractmethod
+    @lru_cache  # noqa: B019
```
Add maxsize=2. I feel 2 is enough. But we can discuss this.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+    ) -> torch.Tensor:
+        return torch.cat(shards, dim=shard_dim)
+
+    @lru_cache  # noqa: B019
```
make sure we set maxsize.

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+            f" but got Q_LEN={Q_LEN} and KV_LEN={KV_LEN}"
+        )
+
+        # check Q_BLOCK_SIZE == KV_BLOCK_SIZE because this should alway be true if Q_LEN == KV_LEN
```
```suggestion
        # check Q_BLOCK_SIZE == KV_BLOCK_SIZE because this should always be true if Q_LEN == KV_LEN
```

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -1456,3 +1574,162 @@ def set_rotate_method(rotate_method: str) -> None:
+        device_type = device_mesh.device_type
+
+        # the sharding function is configurable
+        sharding_map = sharder.gen_block_sharding_map(cp_group_size, block_mask).to(
```
`_rewrite_mask_mod_for_cp()`, `_create_block_mask_cached()`,  `_cp_block_mask`, and `_sharding_map` should be the methods and attributes of `FlashAttentionSharder`.  `ContextParallelMode` should only keep `_sharder`. This can generalize `ContextParallelMode` be used for SDPA as well.

drisspg commented on test/distributed/tensor/test_attention.py:
```diff
@@ -437,5 +439,131 @@ def _test_ring_attention_custom_transformer(self, rotater: _RotateMethod) -> Non
+
+        block_mask = create_block_mask(
+            causal_mask,
+            B=bs,
```
Do we have a different mask for each B and H dim?

drisspg replied:
for this one we dont so you can set to None for BS and H to save some memory

drisspg replied:
yeah 1 or `None`

drisspg commented on test/distributed/tensor/test_attention.py:
```diff
@@ -437,5 +439,131 @@ def _test_ring_attention_custom_transformer(self, rotater: _RotateMethod) -> Non
+        # calling into flex_attention_hop.
+        block_mask_post_sharding = create_block_mask(
+            causal_mask,
+            B=bs,
```
same here

drisspg commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+        return torch.cat(all_buffers, dim=seq_dim)
+
+
+class _RoundRobinLoadBalancer(_LoadBalancer):
```
out of curiosity if  a user knows a better way to shard and unshard for their specific mask is, they only need to subclass the sharder right?

XilunWu replied:
That's right.

drisspg commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+        self, num_chunks: int, block_mask: BlockMask, **kwargs: dict[str, Any]
+    ) -> torch.Tensor:
+        block_mask_tuple = block_mask.as_tuple()
+        assert len(block_mask_tuple) == 13
```
nit: is there a good reason to operate on the tuple as opposed to the named members?

XilunWu replied:
Good catch. I think I only need `as_tuple()` when need to pass the block_mask to HOP call, and can use named members in other places.

XilunWu committed:
Update on "[cp] dispatch flex_attention to CP impl in TorchDispatchMode"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[cp] dispatch flex_attention to CP impl in TorchDispatchMode"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

drisspg commented:
Looks pretty good let me know when ready for another review

XilunWu committed:
Update on "[cp] dispatch flex_attention to CP impl in TorchDispatchMode"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

drisspg commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -62,6 +72,245 @@ class _ContextParallelOptions:
+    @abstractmethod
+    def shard(
+        cls, buffer: torch.Tensor, mesh: DeviceMesh, shard_dim: int
+    ) -> torch.Tensor: ...
```
Nit; maybe some docs on how shard and unshard should be written

fegin reviewed (approved):
Stamp to unblock, please address the comments before landing.

XilunWu committed:
Update on "[cp] dispatch flex_attention to CP impl in TorchDispatchMode"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[cp] dispatch flex_attention to CP impl in TorchDispatchMode"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[cp] dispatch flex_attention to CP impl in TorchDispatchMode"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -227,6 +227,23 @@ def wrapper(*args, **kwargs):
     return decorator
 
 
+def skip_if_lt_world_size():
```
@wanchaol the background is, in this test we want to have a check on `#gpus >= self.world_size` while `skip_if_lt_x_gpu` can only check hard coded constant number (e.g. 4).  Passing `self.world_size` results in an error. `skip_if_lt_world_size` is almost the same as `skip_if_lt_x_gpu(self.world_size)`.

One alternative is `@skip_if_lt_x_gpu(torch.cuda.device_count() if torch.cuda.is_available() else 4)` but it looks quite too cumbersome.

wanchaol replied:
I see, make sense, yeah we can keep the changes here

XilunWu committed:
Update

[ghstack-poisoned]

XilunWu committed:
Update on "[cp] dispatch flex_attention to CP impl in TorchDispatchMode"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[cp] dispatch flex_attention to CP impl in TorchDispatchMode"


## Test
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/debug/_comm_mode.py:
```diff
@@ -733,3 +734,30 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+
+
+# register flex_attention HOP to CommDebugMode
+@flex_attention_hop.py_impl(CommDebugMode)
```
why one need to register a hop here?

wanchaol reviewed (changes requested):
Requesting changes for a few reasons:
* For a relatively complicated changes like this, please update your PR summary with the design so that reviewers can read it, instead of only leaving a test command.
* From API side, I don't think the CP APIs should leak the "Sharder" concept to the users, it is confusing with some other concepts (i.e. the Shard placement). I think it should not be a "Sharder" actually, it is mainly a way to achieve load balancing, the API design needs some further discussion.

fegin commented:
@wanchaol 
> From API side, I don't think the CP APIs should leak the "Sharder" concept to the users, it is confusing with some other concepts (i.e. the Shard placement). I think it should not be a "Sharder" actually, it is mainly a way to achieve load balancing, the API design needs some further discussion.

Regarding the naming issue, we can discuss alternative names if `Sharder` is not suitable. However, it's essential to note that `Sharder` (or any other name) in this context refers specifically to partitioning the input batch along the sequence dimension to achieve load balancing, which is unrelated to DTensor.

To clarify, is the concern related to the naming of `Sharder` or the exposure of the `input partition` concept itself? The input partition needs to be exposed as only users know how to achieve a good load balancing given a customized mask. Though we also plan to do auto load balancing based on the mask, some advanced users have expressed the need to manually control the partition.

An alternative is to configuring `Sharder` an opt-in API that users can call before CP. Thus, most users can just opt to use the built-in feature without worrying the load balancing concept. In such a case, users will just pass in the global `BlockMask`. If users would like to perform its own partitioning, then users can pass the `Sharder` instead of `BlockMask`.

drisspg commented:
I very much believe we want to not rely on auto-load balancing for all users and need way for users to explicit partition

fegin commented:
@wanchaol I agree current CP doesn't work well with DTensor. But I don't agree with you that load balancing should be decoupled from the input sharding. How the input is sharded and arranged will affect the load balancing. Even though we just simply perform allgather on KV, Q is not gathered and how Q is sharded is decided by the input sharding. 

The use case you mentioned is definitely an important one and advanced one. However, a collective (likely to be alltoall) for Q is required for this case, even for ring attention implementation. IMO, we should not sacrifice the performance of the basic use case. I definitely agree this is a direction we should solve, but I will argue that this is the gap of DTensor.

IMO, what @kwen2501 proposed is the right solution to support both with a good UX -- we let DTensor to support a special shard (a sharding that can represent rank0 has the chunk0 and chunk N and rank has the chunk 1 and chunk N -1, ...), I don't know a suitable name,  bit-map-annotation shard?

With this representation, users for the current CP use case can simply partition the input batch using this bit-map-annotation shard to achieve load balancing with extra alltoall. Users can also opt to not shard the input in this way if they have a special use case to do attention and are okay to do alltoall.  Either way, DTensor dispatcher will decide whether to perform alltoall.

fegin commented:
@wanchaol As for dispatcher, we can ask users to wrap SDPA/FlexAttention inside a module and use parallelize_module as we discussed. I don't have a strong opinion about this. cc., @XilunWu
